src/js/node/net.ts

```diff
@@ -667,264 +1122,289 @@ Socket.prototype[kCloseRawConnection] = function () {
-      }
-      tls.requestCert = true;
-      tls.session = session || tls.session;
-      this.servername = tls.servername;
-      tls.checkServerIdentity = checkServerIdentity || tls.checkServerIdentity;
-      this[bunTLSConnectOptions] = tls;
-      if (!connection && tls.socket) {
-        connection = tls.socket;
-      }
+  {
+    const [options, connectListener] =
+      $isArray(args[0]) && args[0][normalizedArgsSymbol] ? args[0] : normalizeArgs(args);
+    let connection = this[ksocket];
+    let upgradeDuplex = false;
+    let { port, host, path, socket, rejectUnauthorized, checkServerIdentity, session, fd, pauseOnConnect } = options;
+    this.servername = options.servername;
+    if (socket) {
+      connection = socket;
     }
-    if (connection) {
-      if (
-        typeof connection !== "object" ||
-        !(connection instanceof Socket) ||
-        typeof connection[bunTlsSymbol] === "function"
-      ) {
-        if (connection instanceof Duplex) {
-          upgradeDuplex = true;
+    if (fd) {
+      Bun.connect({
+        data: this,
+        fd: fd,
+        socket: SocketHandlers,
+        allowHalfOpen: this.allowHalfOpen,
+      }).catch(error => {
+        if (!this.destroyed) {
+          this.emit("error", error);
+          this.emit("close");
+        }
+      });
+    }
+    this.pauseOnConnect = pauseOnConnect;
+    if (!pauseOnConnect) {
+      process.nextTick(() => {
+        this.resume();
+      });
+      this.connecting = true;
+    }
+    if (fd) {
+      return this;
+    }
+    if (
+      // TLSSocket already created a socket and is forwarding it here. This is a private API.
+      !(socket && $isObject(socket) && socket instanceof Duplex) &&
+      // public api for net.Socket.connect
+      port === undefined &&
+      path == null
+    ) {
+      throw $ERR_MISSING_ARGS(["options", "port", "path"]);
+    }
+    const bunTLS = this[bunTlsSymbol];
+    var tls: any | undefined = undefined;
+    if (typeof bunTLS === "function") {
+      tls = bunTLS.$call(this, port, host, true);
+      // Client always request Cert
+      this._requestCert = true;
+      if (tls) {
+        if (typeof rejectUnauthorized !== "undefined") {
+          this._rejectUnauthorized = rejectUnauthorized;
+          tls.rejectUnauthorized = rejectUnauthorized;
         } else {
-          throw new TypeError("socket must be an instance of net.Socket or Duplex");
+          this._rejectUnauthorized = tls.rejectUnauthorized;
+        }
+        tls.requestCert = true;
+        tls.session = session || tls.session;
+        this.servername = tls.servername;
+        tls.checkServerIdentity = checkServerIdentity || tls.checkServerIdentity;
+        this[bunTLSConnectOptions] = tls;
+        if (!connection && tls.socket) {
+          connection = tls.socket;
         }
       }
+      if (connection) {
+        if (
+          typeof connection !== "object" ||
+          !(connection instanceof Socket) ||
```

Comment:
Every one of these instanceof calls can throw. Does Node do instanceof this way?